File: backend/clinprecision-user-service/src/main/java/com/clinprecision/userservice/service/UserTypeService.java
package com.clinprecision.userservice.service;


import com.clinprecision.common.dto.UserTypeDto;
import com.clinprecision.common.entity.UserTypeEntity;

import java.util.List;

/**
 * Service interface for managing user types
 */
public interface UserTypeService {
    
    /**
     * Get all user types
     * @return List of all user types
     */
    List<UserTypeDto> getAllUserTypes();
    
    /**
     * Get user type by ID
     * @param id User type ID
     * @return User type with the specified ID
     */
    UserTypeDto getUserTypeById(Long id);
    
    /**
     * Create a new user type
     * @param userTypeDto User type data
     * @return Created user type
     */
    UserTypeDto createUserType(UserTypeDto userTypeDto);
    
    /**
     * Update an existing user type
     * @param id User type ID
     * @param userTypeDto Updated user type data
     * @return Updated user type
     */
    UserTypeDto updateUserType(Long id, UserTypeDto userTypeDto);
    
    /**
     * Delete a user type
     * @param id User type ID
     */
    void deleteUserType(Long id);
    
    /**
     * Get user types by category
     * @param category User type category
     * @return List of user types with the specified category
     */
    List<UserTypeDto> getUserTypesByCategory(UserTypeEntity.UserCategory category);
    
    /**
     * Get user type by name
     * @param name User type name
     * @return User type with the specified name
     */
    UserTypeDto getUserTypeByName(String name);
    
    /**
     * Get user type by code
     * @param code User type code
     * @return User type with the specified code
     */
    UserTypeDto getUserTypeByCode(String code);
}

File: frontend/clinprecision/src/services/UserTypeService.js
// src/services/UserTypeService.js
import ApiService from './ApiService';

export const UserTypeService = {
  /**
   * Get all user types
   * @returns {Promise} - Promise with user types data
   */
  getAllUserTypes: async () => {
    try {
      const response = await ApiService.get('/users-ws/usertypes');
      return response.data;
    } catch (error) {
      console.error("Error fetching user types:", error);
      throw error;
    }
  },

  /**
   * Get a specific user type by ID
   * @param {string} id - User type ID
   * @returns {Promise} - Promise with user type data
   */
  getUserTypeById: async (id) => {
    try {
      const response = await ApiService.get(`/users-ws/usertypes/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching user type ${id}:`, error);
      throw error;
    }
  },

  /**
   * Create a new user type
   * @param {Object} userTypeData - User type data
   * @returns {Promise} - Promise with created user type data
   */
  createUserType: async (userTypeData) => {
    try {
      const response = await ApiService.post('/users-ws/usertypes', userTypeData);
      return response.data;
    } catch (error) {
      console.error("Error creating user type:", error);
      throw error;
    }
  },

  /**
   * Update an existing user type
   * @param {string} id - User type ID
   * @param {Object} userTypeData - Updated user type data
   * @returns {Promise} - Promise with updated user type data
   */
  updateUserType: async (id, userTypeData) => {
    try {
      const response = await ApiService.put(`/users-ws/usertypes/${id}`, userTypeData);
      return response.data;
    } catch (error) {
      console.error(`Error updating user type ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a user type
   * @param {string} id - User type ID to delete
   * @returns {Promise} - Promise with deletion status
   */
  deleteUserType: async (id) => {
    try {
      const response = await ApiService.delete(`/users-ws/usertypes/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting user type ${id}:`, error);
      throw error;
    }
  }
};

export default UserTypeService;
